File: examples/utilities/PhotosKey.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package javaobject;

import java.util.*;
import java.io.*;
import org.apache.geode.*;
import org.apache.geode.cache.Declarable;
import org.apache.geode.DataSerializer;

import org.apache.geode.logging.internal.log4j.api.LogService;

public class PhotosKey implements DataSerializable {
  public List<String> people;
  public Date rangeStart;
  public Date rangeEnd;

  static {
     Instantiator.register(new Instantiator(javaobject.PhotosKey.class, 500) {
     public DataSerializable newInstance() {
        return new PhotosKey();
     }
   });
  }

  /* public no-arg constructor required for DataSerializable */  
  public PhotosKey() {}

  public PhotosKey(List<String> names, Date start, Date end){
    people = names;
	rangeStart = start;
	rangeEnd = end;
  }
  
  public void fromData(DataInput in) throws IOException, ClassNotFoundException {
    this.people = DataSerializer.readObject(in);
	this.rangeStart = DataSerializer.readDate(in);
	this.rangeEnd = DataSerializer.readDate(in);
  }
  
  public void toData(DataOutput out) throws IOException {
    DataSerializer.writeObject(this.people, out);
    DataSerializer.writeDate(this.rangeStart, out);
    DataSerializer.writeDate(this.rangeEnd, out);
  } 

  public int hashCode() {
	LogService.getLogger().warn("hashCode = {}", Objects.hash(people, rangeStart, rangeEnd));
	return Objects.hash(people, rangeStart, rangeEnd);
  }

  public boolean equals(final Object obj)
  {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    final PhotosKey other = (PhotosKey) obj;

    if (!people.equals(other.people))
      return false;

	if (!rangeStart.equals(other.rangeStart) || !rangeEnd.equals(other.rangeEnd))
	  return false;

    return true;
  }
}

File: examples/dotnet/classaskey/PhotosKey.cs
/*
* Licensed to the Apache Software Foundation (ASF) under one or more
* contributor license agreements.  See the NOTICE file distributed with
* this work for additional information regarding copyright ownership.
* The ASF licenses this file to You under the Apache License, Version 2.0
* (the "License"); you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
using Apache.Geode.Client;
using System;
using System.Collections.Generic;

namespace Apache.Geode.Examples.ClassAsKey
{
  public class PhotosKey : IDataSerializable, ICacheableKey
  {
    public List<String> people;
    public DateTime rangeStart;
    public DateTime rangeEnd;

    // A default constructor is required for deserialization
    public PhotosKey() { }

    public PhotosKey(List<String> names, DateTime start, DateTime end)
    {
      people = names;

      // Geode server defaults to Utc to ensure hashes match between client and
      // server
      TimeZone tz = TimeZone.CurrentTimeZone;
      rangeStart = tz.ToUniversalTime(start);
      rangeEnd = tz.ToUniversalTime(end);
    }

    public override string ToString()
    {
      string result = "{";
      for (int i = 0; i < people.Count; i++)
      {
        result += people[i];
        if (i<people.Count-1)
          result += ", ";
      }
      result += "} from ";
      return result + rangeStart.ToString() + " to " +
        rangeEnd.ToString();
    }

    public void ToData(DataOutput output)
    {
      output.WriteObject(people);
      output.WriteDate(rangeStart);
      output.WriteDate(rangeEnd);
    }

    public void FromData(DataInput input)
    {
      people = (List<String>)input.ReadObject();
      rangeStart = (DateTime)input.ReadDate();
      rangeEnd = (DateTime)input.ReadDate();
    }

    public ulong ObjectSize
    {
      get { return 0; }
    }

    public bool Equals(ICacheableKey other)
    {
      return Equals((object)other);
    }

    public override bool Equals(object obj)
    {
      if (this == obj)
      {
        return true;
      }

      if (GetType() != obj.GetType())
      {
        return false;
      }

      PhotosKey otherKey = (PhotosKey)obj;
      return (people == otherKey.people &&
        rangeStart == otherKey.rangeStart &&
        rangeEnd == otherKey.rangeEnd);
    }

    public override int GetHashCode()
    {
      return Objects.Hash(people, rangeStart, rangeEnd);
    }

    public static ISerializable CreateDeserializable()
    {
      return new PhotosKey();
    }
  }
}

